From 541119ad1afcc9da53a314dd9ebf44ad60735b38 Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Mon, 5 Jan 2026 09:35:37 +0200
Subject: [PATCH] net: phy: realtek: RTL8211F: add LED mode configuration via
 Device Tree
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add support for the Device Tree property realtek,led-modes, which
defines
per-LED mode configuration.
The property is a vector of zero to three 32-bit cells (u32), each
in the range 0x0–0xf, where each cell sets the mode for one LED.
Each cell’s value is a bitmap composed of the following bit flags:
 - 0b0001 - activity
 - 0b0010 - link10
 - 0b0100 - link100
 - 0b1000 - link1000
Unset LEDs are omitted and will use their default configuration.

Signed-off-by: Ilya Ledvich <ilya@compulab.co.il>
Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 drivers/net/phy/realtek.c                  | 56 ++++++++++++++++++++++
 include/dt-bindings/net/realtek-rtl8211x.h | 23 +++++++++
 2 files changed, 79 insertions(+)
 create mode 100644 include/dt-bindings/net/realtek-rtl8211x.h

diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index fb608b31a574..b9582be13732 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -12,6 +12,7 @@
 #include <linux/phy.h>
 #include <linux/module.h>
 #include <linux/delay.h>
+#include <dt-bindings/net/realtek-rtl8211x.h>
 
 #define RTL821x_PHYSR				0x11
 #define RTL821x_PHYSR_DUPLEX			BIT(13)
@@ -38,6 +39,14 @@
 #define RTL8211F_ALDPS_ENABLE			BIT(2)
 #define RTL8211F_ALDPS_XTAL_OFF			BIT(12)
 
+#define RTL8211F_LCR				0x10
+#define RTL8211F_SPEED_10			BIT(0)
+#define RTL8211F_SPEED_100			BIT(1)
+#define RTL8211F_SPEED_1000			BIT(3)
+#define RTL8211F_ACTITITY			BIT(4)
+#define RTL8211F_LED_CTRL			5
+#define RTL8211F_LEDS_NUM			3
+
 #define RTL8211E_CTRL_DELAY			BIT(13)
 #define RTL8211E_TX_DELAY			BIT(12)
 #define RTL8211E_RX_DELAY			BIT(11)
@@ -338,6 +347,46 @@ static int rtl8211c_config_init(struct phy_device *phydev)
 			    CTL1000_ENABLE_MASTER | CTL1000_AS_MASTER);
 }
 
+static int rtl8211f_config_leds(struct phy_device *phydev)
+{
+	struct device *dev = &phydev->mdio.dev;
+	u32 led_modes[RTL8211F_LEDS_NUM];
+	int len;
+	int ret=0;
+
+	len = of_property_read_variable_u32_array(dev->of_node, "realtek,led-modes",
+						  led_modes, 0, ARRAY_SIZE(led_modes));
+	if (len > 0) {
+		u16 reg = 0;
+		int i;
+
+		for (i = 0; i < len; i++) {
+			led_modes[i] &= PHY_LED_MASK;
+			/* RTL8211F: Active makes no sense if no LINK Bit set */
+			if (led_modes[i] == PHY_LED_ACTIVITY)
+				led_modes[i] = PHY_LED_LINK_ACTIVITY;
+
+			/* Parse the requested LED mode and set the corresponding bits in the control register */
+			reg |= (((led_modes[i] & PHY_LED_ACTIVITY)   ? (RTL8211F_ACTITITY)   : (0)) |
+				((led_modes[i] & PHY_LED_SPEED_10)   ? (RTL8211F_SPEED_10)   : (0)) |
+				((led_modes[i] & PHY_LED_SPEED_100)  ? (RTL8211F_SPEED_100)  : (0)) |
+				((led_modes[i] & PHY_LED_SPEED_1000) ? (RTL8211F_SPEED_1000) : (0))) << (i * RTL8211F_LED_CTRL);
+		}
+		dev_info(dev, "Applying LEDs configuration found in Device Tree\n");
+		dev_dbg(dev, "Set LCR (LED Control Register) value: 0x%04x\n", reg);
+		ret = phy_modify_paged_changed(phydev, 0xd04, RTL8211F_LCR, GENMASK(15,0), reg);
+		if (ret < 0) {
+			dev_err(dev, "Failed to update LED Control Register\n");
+			return ret;
+		}
+	} else {
+		dev_info(dev, "No valid LEDs configuration found, use defaults\n");
+		return -ENODATA;
+	}
+
+	return 0;
+}
+
 static int rtl8211f_config_init(struct phy_device *phydev)
 {
 	struct rtl821x_priv *priv = phydev->priv;
@@ -420,6 +469,13 @@ static int rtl8211f_config_init(struct phy_device *phydev)
 		}
 	}
 
+	ret = rtl8211f_config_leds(phydev);
+	if (ret < 0) {
+		dev_warn(dev, "Unable to apply LEDs settings from Device Tree, use defaults\n");
+	}
+	dev_dbg(dev, "Read LCR (LED Control Register): 0x%04x\n",
+		(unsigned int)phy_read_paged(phydev, 0xd04, RTL8211F_LCR));
+
 	return genphy_soft_reset(phydev);
 }
 
diff --git a/include/dt-bindings/net/realtek-rtl8211x.h b/include/dt-bindings/net/realtek-rtl8211x.h
new file mode 100644
index 000000000000..47235bf7d1cf
--- /dev/null
+++ b/include/dt-bindings/net/realtek-rtl8211x.h
@@ -0,0 +1,23 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _DT_BINDINGS_REALTEK_RTL8211X_H
+#define _DT_BINDINGS_REALTEK_RTL8211X_H
+
+/* PHY LED status bitmap flags */
+#define PHY_LED_ACTIVITY	(1 << 0)
+#define PHY_LED_SPEED_10	(1 << 1)
+#define PHY_LED_SPEED_100	(1 << 2)
+#define PHY_LED_SPEED_1000	(1 << 3)
+#define PHY_LED_MASK		0xf
+
+/* Link Up (Any speed) */
+#define PHY_LED_LINK		(PHY_LED_SPEED_10  | \
+				 PHY_LED_SPEED_100 | \
+				 PHY_LED_SPEED_1000)
+
+/* Link Up (Any speed) + Activity (RX/TX) */
+#define PHY_LED_LINK_ACTIVITY	(PHY_LED_LINK | PHY_LED_ACTIVITY)
+/* Special modes */
+#define PHY_LED_OFF		(0)
+#define PHY_LED_DEFAULT		(PHY_LED_OFF)
+
+#endif
-- 
2.34.1

